### PR TITLE
20230320 resolve ipfs links

### PR DESCRIPTION
**Pull request by vadim-frolov**

### CODE REVIEW

File: bin/node/cli/src/main.rs [hunk 1]
```diff
@@ -141,6 +139,15 @@ async fn main() {
         )
         .expect("Failed to create wasm vm");
 
+    if config.ipfs_clients.len() > 0 {
+        resolver = resolver.with_link_resolver(
+            "ipfs".to_string(),
+            Box::new(
+                IpfsLinkResolver::new(&config.ipfs_clients).expect("failed to create IPFS client"),
```
Comment (jorgeantonio21) [inline bin/node/cli/src/main.rs:146]: nit: change the `client` to `clients` in the error message

File: offchain/src/link_resolvers/ipfs.rs [hunk 2]
```diff
@@ -1,18 +1,70 @@
...
+enum IpfsClientState {
+    Online = 0,
+    Timeout = 1,
+}
+#[allow(dead_code)]
+struct IpfsClient {
+    address: String,
+    state: IpfsClientState,
+    last_state_update: Instant,
+    // stats
```
Comment (jorgeantonio21) [inline offchain/src/link_resolvers/ipfs.rs:17]: todo?

File: offchain/src/link_resolvers/ipfs.rs [hunk 3]
```diff
@@ -1,18 +1,70 @@
...
+                    last_state_update: Instant::now(),
+                })
+                .collect(),
+        })
+    }
 
-impl IPFSLinkResolver {
-    pub fn new() -> Result<Self> {
-        Ok(Self {})
+    fn get_best_client(&mut self) -> Option<&IpfsClient> {
```
Comment (jorgeantonio21) [inline offchain/src/link_resolvers/ipfs.rs:40]: How can we evaluate which client is the best ?

**Review by jorgeantonio21 — APPROVED** [review]
LGTM !

Left nit comments/suggestions, feel free to ignore

File: offchain/src/link_resolvers/ipfs.rs (hunk 3 above)
Comment (vadim-frolov) [reply offchain/src/link_resolvers/ipfs.rs:40]: Collect statistics on timeouts, set state to offline, retry after a while, etc. ipfs nodes are not that reliable.

File: offchain/src/link_resolvers/ipfs.rs (hunk 2 above)
Comment (vadim-frolov) [reply offchain/src/link_resolvers/ipfs.rs:17]: yes. will improve on the next iteration. thanks

**Pull request merged**